src/new_index/schema.rs

```diff
@@ -270,6 +269,149 @@ impl Indexer {
         Ok(result)
     }
 
+    /// Clean up orphaned data using the specific list of removed headers
+    /// This is much more efficient than scanning the entire database
+    fn cleanup_orphaned_data(&self, orphaned_headers: &[HeaderEntry]) -> Result<()> {
+        if orphaned_headers.is_empty() {
+            return Ok(());
+        }
+
+        let min_height = orphaned_headers.first().unwrap().height();
+        let max_height = orphaned_headers.last().unwrap().height();
+
+        info!(
+            "Cleaning up orphaned data for {} blocks (heights {} to {})",
+            orphaned_headers.len(),
+            min_height,
+            max_height
+        );
+
+        // Build HashSet of orphaned blockhashes and heights for O(1) lookup
+        let orphaned_hashes: HashSet<BlockHash> = orphaned_headers
+            .iter()
+            .map(|h| *h.hash())
+            .collect();
+
+        let orphaned_heights: HashSet<usize> = orphaned_headers
+            .iter()
+            .map(|h| h.height())
+            .collect();
+
+        self.cleanup_history(&orphaned_heights)?;
+        self.cleanup_confirmations(&orphaned_hashes)?;
```

Comment:
The confirmations index could be removed entirely, see [this comment](https://github.com/Blockstream/electrs/pull/170/files#r2453195339).